In [1]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - 28 weeks of life of the banana plant
---

In [3]:
temperature = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/28-weeks_January-12_August-03_2018/' \
                                 'Temperatura_January-12_August-03_2018.csv', )

In [4]:
print(temperature.shape)
temperature.head()

(9752, 2)


Fecha:  Temperatura (C)
0  2018-01-12 00:17:28            25.06
1  2018-01-12 00:47:28            24.81
2  2018-01-12 01:17:27            24.62
3  2018-01-12 01:47:27            24.56
4  2018-01-12 02:17:30            24.00

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [5]:
print(temperature.isnull().any())
temperature.isnull().values.any()

Fecha:             False
Temperatura (C)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Temperatura (C)` ** and assigning them to the matrix 

`temperature_array` created such as follow:

In [6]:
temperature_array = temperature.iloc[:, 1].values

In [7]:
# Rehape the luminosity_luxes array 
temperature_array = temperature_array.reshape(-1, 1)
temperature_array

array([[25.06],
       [24.81],
       [24.62],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

## 1.3 Generating descriptive data to dataset
---

In [8]:
col = ['Temperatura (C)']
temperature_values_df = pd.DataFrame(temperature_array, columns=col)
temperature_values_df_describe = temperature_values_df.describe()

In [9]:
temperature_values_df_describe

Temperatura (C)
count      9752.000000
mean         27.529622
std           3.731210
min          21.370000
25%          24.560000
50%          26.310000
75%          30.620000
max          40.250000

In [10]:
# Export this descriptive data to comma separated values and java script object notation
temperature_values_df_describe.to_csv('../../../../data/interim/Temperature/28-weeks_January-12_August-03_2018/' +'\n' 
                              'Temperature_Describe_January-12_August-03.cvs', sep=',', header=True, index=True)
temperature_values_df_describe.to_json('../../../../data/interim/Temperature/28-weeks_January-12_August-03_2018/' +'\n'
                                     'Temperature_Describe_January-12_August-03.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`temperature_values_df`** dataset with 9752 samples rows.

In [13]:
print(temperature_values_df.shape)
temperature_values_df.head()

(9752, 1)


Temperatura (C)
0            25.06
1            24.81
2            24.62
3            24.56
4            24.00

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`temperature_values_df`** dataframe to numpy array such as follow:

In [14]:
# numpy_temperature_values = temperature_values_df.reset_index().values
numpy_temperature_values = temperature_values_df.values

In [15]:
numpy_temperature_values

array([[25.06],
       [24.81],
       [24.62],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

We compose the following datasets from **`numpy_temperature_values`** array :

- `temperature_values_train`, which is the training matrix
- `temperature_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_temperature_values`** which contain all column features of luminosity.

Of this way, we pass all data (9750 samples rows) to from them `temperature_values_train`

and `temperature_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **9752 * 0.2 = 1951 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **9752 * 0.8 = 7801  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing temperature datasets
---

- `temperature_values_train`, which is the training matrix
- `temperature_values_test`, the testing matrix

In [18]:
temperature_values_train, temperature_values_test = train_test_split(numpy_temperature_values, test_size = 0.2)

In [19]:
# We have 7801 rows to luminosity_luxes_train
print(type(temperature_values_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , temperature_values_train.shape)
print('\n')

# And we have 1951 rows to luminosity_luxes_test
print(type(temperature_values_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , temperature_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (7801, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (1951, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations



---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---


We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Temperature (C)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Temperature training dataset

In [20]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', temperature_values_train)

Remember our temperature training data 
 [[29.81]
 [32.12]
 [23.43]
 ...
 [29.12]
 [25.12]
 [25.75]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `temperature_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [22]:
minmax_scale_training = scaler.fit(temperature_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `temperature_values_train` data are scaled between **0 to 1**  selected range 


In [23]:
# transform precipitations_train data to maximum and mínimum scale value. 
temperature_minmax_training = minmax_scale_training.transform(temperature_values_train)

In [24]:
print ("And, these are our scaled data: " + '\n')
temperature_minmax_training

And, these are our scaled data: 



array([[0.4470339 ],
       [0.56938559],
       [0.10911017],
       ...,
       [0.41048729],
       [0.19862288],
       [0.23199153]])

In [25]:
print('Temperature Training dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_minmax_training[:,0].min()))

print('Temperature Training dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_minmax_training[:,0].max()))

Temperature Training dataset. Minimum value after MaxMinScaler:
Luminosity=0.0
Temperature Training dataset. Maximum value after MaxMinScaler:
Luminosity=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_minmax_training` numpy array

- We export this array to comma separated values 


In [26]:
temperature_train_df = pd.DataFrame(temperature_minmax_training, columns=col)

In [27]:
temperature_train_df

Temperatura (C)
0            0.447034
1            0.569386
2            0.109110
3            0.576271
4            0.271716
5            0.026483
6            0.198623
7            0.668962
8            0.506356
9            0.115996
10           0.129237
11           0.182203
12           0.182203
13           0.105932
14           0.129237
15           0.145657
16           0.082627
17           0.218750
18           0.162076
19           0.139301
20           0.327860
21           0.126059
22           0.195445
23           0.364407
24           0.532839
25           0.387182
26           0.807733
27           0.351165
28           0.185381
29           0.201801
...               ...
7771         0.728284
7772         0.082627
7773         0.119174
7774         0.211864
7775         0.245233
7776         0.248411
7777         0.254767
7778         0.377648
7779         0.320975
7780         0.268008
7781         0.254767
7782         0.211864
7783         0.201801
7784         0.791314
7785         0.095869
7786         0.510064
7787         0.394068
7788         0.105932
7789         0.188559
7790         0.351165
7791         0.595869
7792         0.655720
7793         0.231992
7794         0.274894
7795         0.529661
7796         0.126059
7797         0.188559
7798         0.410487
7799         0.198623
7800         0.231992

[7801 rows x 1 columns]

In [ ]:
In this way we have the dataset `temperature_train_df` standardized and training, and export it to a .csv file

In [29]:
temperature_train_df.to_csv('../../../../data/processed/Temperature/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TRAINING_January-12_August-03.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [31]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', temperature_values_test)

Remember our precipitations testing data 
 [[29.5 ]
 [22.18]
 [24.12]
 ...
 [28.87]
 [25.62]
 [31.37]]


In [32]:
minmax_scale_test = scaler.fit(temperature_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
temperature_minmax_test = minmax_scale_test.transform(temperature_values_test)

In [33]:
print ("And, these are our testing scaled data: " + '\n')
temperature_minmax_test

And, these are our testing scaled data: 



array([[0.43777778],
       [0.03111111],
       [0.13888889],
       ...,
       [0.40277778],
       [0.22222222],
       [0.54166667]])

In [34]:
print('Temperature Testing dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_minmax_test[:,0].min()))

print('Temperature Testing dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_minmax_test[:,0].max()))

Temperature Testing dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Luminosity Testing dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_minmax_test` numpy array

- We export this array to comma separated values

In [35]:
temperature_test_df = pd.DataFrame(temperature_minmax_test, columns=col)

In [36]:
print(temperature_test_df.shape)
temperature_test_df.head()

(1951, 1)


Temperatura (C)
0         0.437778
1         0.031111
2         0.138889
3         0.180556
4         0.635556

In [37]:
temperature_test_df.to_csv('../../../../data/processed/Temperature/28-weeks_January-12_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TESTING_January-12_August-03.csv', sep=',', header=True, index=False)